core/src/main/java/org/apache/calcite/sql/SqlNumericLiteral.java

```diff
@@ -99,17 +99,23 @@ public boolean isExact() {
     if (isExact) {
       int scaleValue = requireNonNull(scale, "scale");
       if (0 == scaleValue) {
-        BigDecimal bd = getValueNonNull();
-        SqlTypeName result;
-        long l = bd.longValue();
-        if ((l >= Integer.MIN_VALUE) && (l <= Integer.MAX_VALUE)) {
-          result = SqlTypeName.INTEGER;
-        } else {
-          result = SqlTypeName.BIGINT;
+        try {
+          BigDecimal bd = getValueNonNull();
```

Comment:
numeric literals are always represented as Decimal by the parser.
Here we check whether the value can be represented by a narrower type without information loss.